text/0000-pub-restricted.md

```diff
@@ -0,0 +1,924 @@
+
+[when public (RFC 136)]: https://github.com/rust-lang/rfcs/blob/master/text/0136-no-privates-in-public.md#when-is-an-item-public
+
+However, since this RFC is introducing multiple kinds of `pub`, we
+should address the topic of what *is* the `pub`-ness of associated
+items.
+
+ * When analyzing a trait definition, then associated items should be
+   considered to inherit the `pub`-ness, if any, of their defining
+   trait.
+
+   We want to make sure that this code continues to work:
+
+   ```rust
+   mod a {
+       struct S(String);
+       trait Trait {
+           fn make_s(&self) -> S; // referencing `S` is ok, b/c `Trait` is not `pub`
+       }
+   }
+   ```
+
+   And under this RFC, we now allow this as well:
+
+   ```rust
+   mod a {
+       struct S(String);
+       mod b {
+           pub(a) trait Trait {
+               fn mk_s(&self) -> ::a::S;
+               // referencing `::a::S` is ok, b/c `Trait` is restricted to `::a`
+           }
+       }
+       use self::b::Trait;
+   }
+   ```
+
+   Note that in stable Rust today, it is an error to declare the latter trait
+   within `mod b` as non-`pub` (since the `use self::b::Trait` would be
+   referencing a private item),
+   *and* in the Rust nightly channel it is a warning to declare it
+   as `pub trait Trait { ... }`.
+
+   The point of this RFC is to give users a sensible way to declare
+   such traits within `b`, without allowing them to be exposed outside
+   of `a`.
+
+ * When analyzing an `impl Trait for Type`, there may be distinct
+   restrictions assigned to the `Trait` and the `Type`. However,
+   since both the `Trait` and the `Type` must be visible in the
+   context of the module where the `impl` occurs, there should
+   be a subtree relationship between the two restrictions; in other
+   words, one restriction should be less than (or equal to) the other.
+
+   So just use the minimum of the two restrictions when analyzing
+   the right-hand sides of the associated items in the impl.
+
+   Note: I am largely adopting this rule in an attempt to be
+   consistent with [RFC 136][when public (RFC 136)]. I invite
+   discussion of whether this rule actually makes sense as phrased
+   here.
+
+## More examples!
+[examples]: #more-examples
+
+These examples meant to explore the syntax a bit. They are *not* meant
+to provide motivation for the feature (i.e. I am not claiming that the
+feature is making this code cleaner or easier to reason about).
+
+### Impl item example
+[impl item example]: #impl-item-example
+
+```rust
+pub struct S;
+
+mod a {
+    pub fn call_foo(s: &S) { s.foo(); }
+
+    impl S {
+        pub(a) fn foo(&self) { println!("only callable within `a`"); }
+    }
+}
+
+fn rejected(s: &S) {
+    s.foo(); //~ ERROR: `S::foo` not visible outside of module `a`
+}
+```
+
+(You may be wondering: "Could we move that `impl S` out to the
+top-level, out of `mod a`?" Well ... see discussion in the
+[unresolved questions][def-outside-restriction].)
```

Comment:
I don't find this to be a convincing example. You could just leave out the `pub(a)` and the [example would work right now just as expected](http://is.gd/xqoj7z). Adding impls inside submodules is imo the correct way to say it's only accessible in there. The example should probably be showing two submodules accessing the same function. But then it doesn't fit here, but to to the unresolved question point.